### PR TITLE
Fix duplicate email constraint in WP user sync

### PR DESCRIPTION
**Pull request by bruberries**
## Description

When a WordPress user changes their email address to one already held by another MailPoet subscriber (e.g. created during a WooCommerce guest checkout), `WP::synchronizeUser` would fail with a unique constraint violation on `flush()`.

This adds a check before updating the subscriber email: if another subscriber already holds the target email, the duplicate is removed first.

## Code review notes

The duplicate subscriber is removed (not merged) because it typically represents a guest checkout record with no WP user association. Merging subscriber data would add complexity for a rare edge case where the guest record usually has minimal data.

Deletion goes through `SubscribersRepository::bulkDelete()`, which only removes subscribers that are **not** WP users and **not** WooCommerce customers. If the duplicate is a WP user or WooCommerce customer, the email update is safely skipped and a warning is logged.

### Scenario A — duplicate is a regular subscriber (email update succeeds, duplicate removed)

1. Register a WP user with email A.
2. Create a regular MailPoet subscriber with email B (e.g. via a MailPoet subscription form, or **MailPoet → Subscribers → Add New**). Make sure this subscriber is **not** a WP user and **not** a WooCommerce customer.
3. Go to **WP admin → Users**, edit the user, and change their email from A to B.
4. Verify:
   - No error occurs.
   - The MailPoet subscriber list shows only one entry for email B, and it is linked to the WP user (the standalone email-B subscriber has been removed).

### Scenario B — duplicate is a WP user or WooCommerce customer (email update safely skipped)

1. Register a WP user with email A.
2. Place a WooCommerce order as a guest using email B (creates a MailPoet subscriber with `is_woocommerce_user = true`). Alternatively, register a second WP user with email B.
3. Go to **WP admin → Users**, edit the first user, and change their email from A to B.
4. Verify:
   - No unique-constraint error / fatal error occurs.
   - The WooCommerce/WP-user subscriber for email B is **preserved** (not deleted).
   - The MailPoet subscriber for the WP user keeps email A (the sync is skipped).
   - A warning is logged to `mailpoet.log` along the lines of *"Cannot update subscriber email: duplicate subscriber is a WP user or WooCommerce customer"*.

## Linked PRs

_N/A_

## Linked tickets

[STOMAIL-7912](https://linear.app/a8c/issue/STOMAIL-7912)

## After-merge notes

_N/A_

## Screenshots or screencast

_N/A_

## Tasks

- [x] I have added a changelog entry
- [x] I followed best practices for translations
- [x] I added sufficient test coverage
- [ ] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:fix/STOMAIL-7912-duplicate-email-constraint)

_The latest successful build from `fix/STOMAIL-7912-duplicate-email-constraint` will be used. If none is available, the link won't work._

### CODE REVIEW

**Comment by bruberries** [comment]
Tagging @zhongruige and @Luc45 as first responders for review.

**Review by yuliyan — COMMENTED** [review]
Tested the change and it works correctly.

However, the removal of a duplicate subscriber and the subsequent update are not atomic - if the update fails, the removed subscriber will be lost. To avoid that, you can wrap both operations in `$this->entityManager->wrapInTransaction` to ensures the delete is rolled back on failure.

Also noticed that the test that you've added is failing. Could you take look and fix it, or let me know if you need any assistance with that.

**Comment by bruberries** [comment]
@coderabbitai review

File: mailpoet/lib/Segments/WP.php [hunk 1]
```diff
@@ -182,10 +182,35 @@ private function handleCreatingOrUpdatingSubscriber(string $currentFilter, \WP_U
...
+    // Uses a DBAL-level transaction (not EntityManager::wrapInTransaction) to
+    // keep the delete and update atomic without risking a permanently closed
+    // EntityManager or premature postRemove lifecycle events.
+    if ($subscriber !== null && $subscriber->getEmail() !== $data['email']) {
+      $this->databaseConnection->beginTransaction();
+      try {
+        $this->databaseConnection->executeStatement(
+          "DELETE FROM {$this->subscribersTable} WHERE email = ? AND id != ?",
+          [$data['email'], $subscriber->getId()]
+        );
```
[the remark this answers is not in the document]
Comment (bruberries) [reply mailpoet/lib/Segments/WP.php]: Fixed it!

**Comment by bruberries** [comment]
Hi @yuliyan, I've addressed your feedback from the earlier review:

1. **Transaction safety** — switched to `bulkDelete()` which properly cleans up related data (segments, tags, custom fields) and restricts deletion to safe duplicates (non-WP, non-WooCommerce subscribers).
2. **Failing test** — fixed the cache issue and added an edge-case test for protected (WooCommerce) subscribers.

Also rebased onto trunk to pick up the latest changes. Could you take another look when you get a chance?

**Review by yuliyan — APPROVED** [review]
@bruberries updated code looks good and works as expected.

Before assigning this for QA testing, could you please update the "QA notes"? Currently, they state that the "email B" user is created through the WooCommerce checkout, which results in a "WooCommerce" subscriber entry. However, according to the code, updates are skipped for WooCommerce customer and WP User subscribers, and duplicate entries are not removed.

**Comment by bruberries** [comment]
@yuliyan Thanks for flagging this! I edited the QA notes now.

**Comment by yuliyan** [comment]
@bruberries, thanks!

@Aschepikov, could you please test this change and verify it works correctly and doesn't introduce any regression. Thanks!

**Comment by Aschepikov** [comment]
@yuliyan on it

**Review by Aschepikov — APPROVED** [review]
Both scenarios were tested. GJ!
